### ISSUE TITLE
Wrong redirect using the GrandID-BankID on mobile devices.

### ISSUE THREAD

**Issue by JCOlofsson** [ActiveLogin/ActiveLogin.Authentication#83]
Thanks for making a great effort to streamline the implementation of BankID! 

We have stumbled upon an unexpected behaviour in the login flow using ActiveLogin with the GrandID BankID implementation in our Identity Server project.

**UNEXPECTED BEHAVIOUR**
BankID application redirects the user to the wrong browser (always default system browser) if the BankID app is opened via the GrandID login page from a native app (IABT) OR non default system web browser (e.g. Chrome on iOS). This unexpected behaviour occurs on Android as well as iOS.

**REPRODUCTION**
0) The device being used to reproduce this behaviour must have a working BankID installed.
1) Open one of the two live project examples (e.g. https://al-samples-mvcclient.azurewebsites.net) on an iOS device using the Chrome web browser OR other non-default web browser (not Safari).
2) Choose to login with one of the following options:
2.a) "GrandID SameDevice" or
2.b) "GrandID OtherDevice". Using this option, be sure to input social security number (SSN) AND launch the BankID application by pressing the button (STARTA BANKID PÅ DEN HÄR ENHETEN) shown on the GrandID login page.
3) Approve (sign) the login request inside the BankID application.
4) This is where the unexpected behaviour happens: Safari opens (system browser) and the user is taken away from the initial application (Chrome browser  in this case) where the user initiated the login flow. The same unexpected behaviour happens if the login flow is performed from a native app (non-browser) using the Safari View Controller (in app browser tab) which now is required by Apples Design Guidelines when logging in via an external login page (e.g Identity Server login page). 

**POSSIBLE SOLUTION**
Apologies if I’m barking up the wrong tree. I’m not as well versed in C# .NET as I am in JS. Perhaps there’s a way to pass a redirectURL to the GrandID API that will instruct the BankID application to open the initial application once signing is complete? E.g. using custom URL schemes, returnURL=se.activesolution.AnAwesomeApp:// to re-open the initial application to complete the login flow?

Happy to contribute on this!
/Christofer

**Comment by fredrik-lundin** [comment]
Hi Christofer,

Well spotted and thanks a lot for taking the time to post such a detailed issue! 

I can reproduce this really easily on an android phone by:

0. (having chrome as my default browser)
1. Open another browser than chrome and navigate to https://al-samples-mvcclient.azurewebsites.net
2. Press "GrandID (SameDevice)"
3. Login as normal following the instructions 
4. When I'm indeed redirected back to Chrome, and an error occurs

We are happily accepting PRs if that's what you meant with contributing. If not, any help such as reviewing or giving input to whoever decides to create a PR for this is also much appreciated! 

Thanks again!

**Comment by JCOlofsson** [comment]
Hi Fredrik,
Thanks for the quick reply.
I'm still on the steep end of the learning curve getting familiar with Identity Server and ActiveLogin.Authentication. I think I'll add most value by testing  and discuss PRs (browsers and native apps) at the moment.

I'll try to get up to speed on Identity Server and ActiveLogin.Authentication meanwhile if someone else would be kind enough to work on a PR.

Cheers

**Comment by PeterOrneholm** [comment]
First of all, what a fantastic structured issue @JCOlofsson! It's a joy being an open source maintainer when the issues looks like this!

And second, thanks for finding and reporting on this!

I will look deeper into this, but here is my initial thoughts on what the issue might be.

First of all, when using GrandID, it's technically GrandID (Svensk E-identitet) that has implemented the redirect flow, we are just builing a layer on top of them. But they can also have an issue of course, and we have good connections to report that. And, because we have implemented BankID ourselfes, I have a feeling for what it can be.

On iOS BankID requires you to provide a returnUrl to the BankID app if you use the "open the app automatically"-flow. Once the login is done, it will tell iOS to launch that URL, and because only Safari can be the default browser, this issue occurs. And because it (for security reasons) requires you to finish the login in the same browser you started it, it will fail.

On the other hand, on Android, it's different. There you are supposed to leave the returnUrl (technically setting it to the string "null"), and the Android app will then fall into the behavior of launching the previously opened app, which should be whatever browser you are using.

These requirements a re described in [section 3 in the BankID docs](https://www.bankid.com/assets/bankid/rp/bankid-relying-party-guidelines-v3.2.1.pdf).

I will notify GrandID about this issue on Android and ask them to look into it.

On iOS on the other hand, I'm not sure if it's actually solvable with "On this device", due to how the auto launch works. I'd say you need to use "On other device" and manually launch the BankID app in this case. Would be great to have any input if you have more info / cases when this works.

**Comment by JCOlofsson** [comment]
Hi Peter! 
Thank you for your quick feedback! 
Very interesting thoughts. Nice to see the returnUrl in the BankID docs. Wasn't sure exactly what was passed to the BankID-app. Feels like the returnURL is key to this issue. 

I've been trying to formulate a well-thought-out reply. Safe to say I'm still churning away over here with my 1-core humanoid brain. I'll have a couple of ideas and I'll try to post a suggestion later tonight.

**Comment by JCOlofsson** [comment]
Ok here goes (apologies for a lengthy reply). 
I've been typing and scratching for a while now. I was initially focused on customising the returnUrl parameter passed from the WebApp / Native App that initiates the login flow. But that won't work all the way as we need to use the https:// scheme when directing to a different page within the same browser window.

I made a table/chart to structure my thoughts around the problem, what's required for the different login scenarios etc. (see below). Pondering on it for a while, it all seem to boil down to the _return_-parameter that is passed on from GrandID to the BankID-app.

### The short version
I believe this issue could be solved if we could append the required URL-scheme (associated with the  Browser or In-App Browser) in the request to GrandID. That way GrandID could compose a URL that will open the correct app from the BankID-app. In other words:  We would hint to GrandID in its request to the BankID-app that it wants a return url parameter with the value googlechrome://login.grandid.se (normally https://login.grandid.se).

### The long version
Each major browser has a custom URL scheme that it is registered to. E.g. entering  googlechrome://www.activesolution.se in Safari in iOS will open the site in Chrome. Same with goes for Firefox (firefox://open-url?url=https://www.activesolution.se). These are the urls the BankID-app should be instructed to use. Same thing if it is a Native app. In that case the hint should state a URL scheme registered with the Native App. E.g. se.activesolution.app://.
GrandID has trouble knowing which application (Browser and In-App Browser) is actually running the GrandID web application (hibernating till the user returns). As soon as the user has signed within the BankID-App the user just has to re-open the application holding the hibernating GrandID-web-app to continue the login flow. At the moment (and I'm guessing here) GrandID is always appending a _return_-parameter value to the BankID-application with the https:// URL scheme, which always results in a redirect to the default browser. 

To prevent this we could pass a "hint" parameter to GrandID so that the GrandID web service could construct the correct return URL. I've tried to illustrate the "hint"-parameter as "customURLhint" (red) in the table below. Implementing such a parameter hands over the responsibility of correctly identifying associated custom URL schemes to the app initiating the login flow.

**Determining which URL scheme to use in the web-app**
WebApp developers could use the user agent string available in the window object to determine Operating System and Browser. An even nicer way to do so is to use the [detect-browser](https://www.npmjs.com/package/detect-browser) library which does this for us.

Very interested to hear your thoughts. Good or bad? Or simply madness from a compliance perspective?



_**Note:** the login scenarios in the chart below illustrate:_
- _Web app authentication running in one of the three major browsers on mobile devices and_ 
- _Native app authentication from an In-App Browser Window (recommended UX by Apple Guidelines)._ 
_These scenarios are (should) be the same on Android. The scenarios exclude Desktop browsers as those scenarios work like a charm._ 

![image](https://user-images.githubusercontent.com/10133791/52652579-aeb98a00-2eee-11e9-8a0f-a2d90ba162b9.png)

**Comment by PeterOrneholm** [comment]
Thanks for a very detailed and thought through @JCOlofsson!

On Android, returnUrl should be set to null, documented in "3.1.2 Parameters in the Start URL" of the [BankID documentation](https://www.bankid.com/assets/bankid/rp/bankid-relying-party-guidelines-v3.2.1.pdf). When doing that, it will automatically launch the previously used app, so it would handle Chrome, Firefox or even custom apps.

I just found [this link](https://developer.chrome.com/multidevice/ios/links) for Chrome on iOS, making it googlechrome**s**:// will also ensure https://. The equivalent documentation for Firefox {can be found here](https://github.com/mozilla-mobile/firefox-ios-open-in-client).

Our own implementation does set redirectUrl to null on Android, so we handle that scenario, but we do not handle googlechromes://, firefox:// or any other special cases on iOS. That would be a good thing to add. I've noticed that Kivra supports googlechromes:// but have not seen anyone else that does this. I'll add a separate issue on this.

But, for your case because you are using GrandID, it's a little bit out of our hands as our implementation "only" wraps their API and sends the user to their site. It would be them that will have to implement such features. Setting null for Android should be the easiest, but the iOS scenarios could probably be handled as well.

Do you want me to reach out to them and point to this issue? Or maybe you could, as a customer of them, do so? If sending them an email, please keep me in the loop if I can assist in any way.

**Comment by PeterOrneholm** [comment]
I forgot to ask, but do we agree that this is not an bug in Active Login, but rather a feature that GrandID needs to implement? And if, after they have fixed it, would require us to adopt to any new API, I think that should be a new Issue.

So, if you agree, I'll close this Issue for now, especially as we moved the part for iOS on our own implementation into issue #86.

**Comment by JCOlofsson** [comment]
Thank you! googlechrome**s**:// was a true aha-moment :). 

Definately, you are so right. This issue should be with GrandID.
I'll get in touch with them and keep you in the loop. Thank you for your guidance!

**Comment by snios** [comment]
@JCOlofsson  Is there an open issue i can track with grandid for this case?

**Comment by JCOlofsson** [comment]
Hi @snios, not that I'm aware. Back then, GrandID's customer support said it was a known limitation in their product and they had an issue in their backlog but they couldn't give me details on when the issue was expected to be resolved. This issue (#83) concerned what Svensk E-Identitet called GrandID v2. Since then Svensk E-Identitet has released at least one new major API version.
Are you experiencing the same issue in their latest API version?

**Comment by snios** [comment]
> Hi @snios, not that I'm aware. Back then, GrandID's customer support said it was a known limitation in their product and they had an issue in their backlog but they couldn't give me details on when the issue was expected to be resolved. This issue (#83) concerned what Svensk E-Identitet called GrandID v2. Since then Svensk E-Identitet has released at least one new major API version.
> Are you experiencing the same issue in their latest API version?

@JCOlofsson  Thanks for the reply. Yes we are experiencing the same Issue with our environment that consists of React Native -> Identity Server 4 -> ActiveLogin with GrandID. 

However we just found out that GrandID accepts a parameter for "appRedirect" and after setting that in our config for GrandID in our Identity project the app launches even tough it's not the smoothest flow. "Hacky, but it works" solution.

**Comment by PeterOrneholm** [comment]
I'm not working activly with the GrandID API anymore - but please file an issue if there are missing API:s or things to be fixed :)

**Comment by PeterOrneholm** [comment]
Hi @snios, @Max-Eriksson and @nios87!

We are not working that much on the GrandID-client anymore (focusing on the native BankID one), so little out of the loop here.
Is there a need that you now need to solve by cloning the repo and compiling the client yourself? Can we add that to the codebase? Any suggestion on what to add?

**Comment by nios87** [comment]
Hey @PeterOrneholm  There is a parameter/config i was going to do a PR for but got unsecure on how or where to add it as there are a few interfaces in the project.  But in essence what i need to add to the postData is appRedirect as above and it would be nice if that where something you could set in the config of "AddGrandId". it should be optional tough and i dont know what default value it should have or if it should just be omitted in the postData if it is not configured.
Im also unsure if it should be in the options part or somewhere else in the AddGrandId builder.

example
![image](https://user-images.githubusercontent.com/89184137/139655452-86237c21-ce31-4ed9-a495-4edd45db3c2d.png)

**Comment by PeterOrneholm** [comment]
Thanks @nios87! I've been trying to read up on the docs now.

As I've understood it - if `appRedirect` is supplied the BankID app will redirect back to that exact URL and not to the Svensk E-identitet URL. It still works because in the app, the Svensk E-identitet page is "open" and will handle the completion if the auth flow? Is this correct?

We don't add new features to the GrandID-things (exception is adding missing paremeters in the API wrapper). So, that we could do (can you do a PR for this)? But we will not do it all the way "to the top". If we were to implement it, I'd like to follow the pattern from the native BankID lib and allow for different values per request (as a OIDC server might serve multiple clients iOS/Android/Web). And the effort to implement, test and maintain this is just too big now.

BUT, you could still use the NuGet package and just override a few classes. By Implementing your own class overriding `GrandIdBankIdHandler` and changing how `GetRedirectUrlAsync` is handled you should be able to accomplish this I think.
